### PR TITLE
Evolving from a saved in-vitro state

### PR DESCRIPTION
**Pull request by geralddejong**
This was needed in order to practically evolve the genes to [turn after completing the first leg](https://youtu.be/If4fDzoAEGg).

### CODE REVIEW

File: client/src/body/fabric-exports.ts [hunk 1]
```diff
@@ -12,22 +12,24 @@ export enum Direction {
...
+//         case Direction.REVERSE:
+//             return right ? Direction.LEFT : Direction.RIGHT
+//         case Direction.LEFT:
+//             return right ? Direction.FORWARD : Direction.REVERSE
+//         case Direction.RIGHT:
+//             return right ? Direction.REVERSE : Direction.FORWARD
+//         default:
+//             return Direction.FORWARD
+//     }
+// }
```
Comment (mitchelljustin) [inline client/src/body/fabric-exports.ts]: Remove commented out code pls

File: client/src/body/fabric-kernel.ts [hunk 2]
```diff
@@ -56,37 +57,59 @@ function createOffsets(faceCountMax: number, fabricBytes: number): IOffsets {
...
+            this.instanceUsed.push(false)
         }
     }
 
-    public get instance(): IFabricInstanceExports[] {
-        return this.instanceArray
+    public createGotchiSeed(location: Vector3, rotation: number, genome: Genome): Gotchi | undefined {
+        const newInstance = this.allocateInstance()
+        if (!newInstance) {
+            return undefined
```
Comment (mitchelljustin) [inline client/src/body/fabric-kernel.ts:99]: Shouldn't you use `null` as a sentinel value instead?

File: client/src/body/fabric-kernel.ts [hunk 3]
```diff
@@ -56,37 +57,59 @@ function createOffsets(faceCountMax: number, fabricBytes: number): IOffsets {
...
         }
         for (let index = 0; index < dimensions.instanceMax; index++) {
-            this.instanceArray.push(new InstanceExports(this.arrayBuffer, this.offsets, exports, dimensions, index))
+            this.instanceArray.push(new InstanceExports(
+                this.arrayBuffer,
+                createOffsets(dimensions.faceCountMax, HEXALOT_SIZE + index * fabricBytes),
+                exports,
+                dimensions,
+                index,
+                toFree => this.instanceUsed[toFree] = false,
```
Comment (mitchelljustin) [inline client/src/body/fabric-kernel.ts:90]: Passing a callback to a constructor is a weird pattern.

File: client/src/gotchi/evolution.ts [hunk 4]
```diff
@@ -21,124 +19,174 @@ const SURVIVAL_RATE = 0.66
...
-            .map(evolver => evolver.gotchi.fabric.vectors)
-            .reduce((prev, array) => {
-                prev.x += array[0]
-                prev.y += array[1]
-                prev.z += array[2]
-                return prev
-            }, new Vector3())
-            .multiplyScalar(1 / evolvers.length)
+        const vectors = evolvers.map(evolver => evolver.vectors)
+        const sumFromArray = (prev: Vector3, array: Float32Array) => {
```
Comment (mitchelljustin) [inline client/src/gotchi/evolution.ts]: A lil sugar
```suggestion
        const sumFromArray = (prev: Vector3, [x,y,z]: Float32Array) => {
```

File: client/src/island/journey.ts [hunk 5]
```diff
@@ -1,28 +1,42 @@
...
+            return undefined
+        }
+        const goTo = this.journey.hexalots[nextHexalot]
+        return new Leg(this.journey, nextHexalot, goTo)
+    }
 }
 
 export class Journey {
-    constructor(private hexalots: Hexalot[]) {
+    constructor(public hexalots: Hexalot[]) {
```
Comment (mitchelljustin) [inline client/src/island/journey.ts]: Just make er `readonly`

File: tslint.json [hunk 6]
```diff
@@ -69,6 +69,7 @@
       "property-declaration"
     ],
     "no-empty": false,
-    "max-line-length": [true, 160]
+    "max-line-length": [true, 160],
+    "no-debugger": false
```
Comment (mitchelljustin) [inline tslint.json:73]: Nice

**Review by mitchelljustin — APPROVED** [review]

File: client/src/body/fabric-kernel.ts (hunk 2 above)
Comment (geralddejong) [reply client/src/body/fabric-kernel.ts:99]: I have no reason, and you don't give one.

File: client/src/body/fabric-kernel.ts (hunk 3 above)
Comment (geralddejong) [reply client/src/body/fabric-kernel.ts:90]: Why?

File: client/src/gotchi/evolution.ts (hunk 4 above)
Comment (geralddejong) [reply client/src/gotchi/evolution.ts]: Nice

**Pull request merged**